src/parameterized_tendencies/gravity_wave_drag/non_orographic_gravity_wave.jl

```diff
@@ -201,252 +196,543 @@ function non_orographic_gravity_wave_tendency!(
+
+    v_endlevel = Fields.level(ᶜv, Spaces.nlevels(axes(ᶜv)))
+    v_endlevel_m1 = Fields.level(ᶜv, Spaces.nlevels(axes(ᶜv)) - 1)
+    Boundary_value = Fields.Field(
+        FT(2) .* Fields.field_values(v_endlevel) .-
+        Fields.field_values(v_endlevel_m1),
+        axes(v_endlevel),
+    )
+    ᶜv_p1 .= field_shiftlevel_up(ᶜv, Boundary_value)
+
+    Boundary_value = Fields.level(ᶜbf, Spaces.nlevels(axes(ᶜbf)))
+    ᶜbf_p1 .= field_shiftlevel_up(ᶜbf, Boundary_value)
+
+    z_endlevel = Fields.level(ᶜz, Spaces.nlevels(axes(ᶜz)))
+    z_endlevel_m1 = Fields.level(ᶜz, Spaces.nlevels(axes(ᶜz)) - 1)
+    Boundary_value = Fields.Field(
+        FT(2) .* Fields.field_values(z_endlevel) .-
+        Fields.field_values(z_endlevel_m1),
+        axes(z_endlevel),
+    )
+    ᶜz_p1 .= field_shiftlevel_up(ᶜz, Boundary_value)
+
+    mask = BitVector(ones(nc))
+    level_end = Spaces.nlevels(axes(ᶜρ))
+    B1 = ntuple(i -> 0.0, Val(nc))
+
+    for ink in 1:gw_nk
+
+        input_u = Base.Broadcast.broadcasted(
+            tuple,
+            ᶜu_p1,
+            ᶜu_source,
+            ᶜbf_p1,
+            ᶜρ,
+            ᶜρ_p1,
+            ᶜρ_source,
+            ᶜz_p1,
+            ᶜz,
+            source_level,
+            gw_Bw,
+            gw_Bn,
+            gw_cw,
+            gw_cn,
+            gw_c,
+            gw_flag,
+            ᶜlevel,
+            gw_source_ampl,
+        )
+
+
+        input_v = Base.Broadcast.broadcasted(
+            tuple,
+            ᶜv_p1,
+            ᶜv_source,
+            ᶜbf_p1,
+            ᶜρ,
+            ᶜρ_p1,
+            ᶜρ_source,
+            ᶜz_p1,
+            ᶜz,
+            source_level,
+            gw_Bw,
+            gw_Bn,
+            gw_cw,
+            gw_cn,
+            gw_c,
+            gw_flag,
+            ᶜlevel,
+            gw_source_ampl,
+        )
+
+        Operators.column_accumulate!(
+            u_waveforcing,
+            input_u;
+            init = (FT(0.0), mask, 0.0, B1),
+            transform = first,
+        ) do (wave_forcing, mask, Bsum, B0),
+        (
+            u_kp1,
+            u_source,
+            bf_kp1,
+            ρ_k,
+            ρ_kp1,
+            ρ_source,
+            z_kp1,
+            z_k,
+            source_level,
+            Bw,
+            Bn,
+            cw,
+            cn,
+            c,
+            flag,
+            level,
+            source_ampl,
+        )
+            if level >= (source_level - 1)
+                FT1 = typeof(u_kp1)
+                kwv = 2.0 * π / ((30.0 * (10.0^ink)) * 1.e3)
+                k2 = kwv * kwv
+                fac = FT1(0.5) * (ρ_kp1 / ρ_source) * kwv / bf_kp1
+                Hb = (z_kp1 - z_k) / log(ρ_k / ρ_kp1) # density scale height
+                alp2 = FT1(0.25) / (Hb * Hb)
+                ω_r = sqrt((bf_kp1 * bf_kp1 * k2) / (k2 + alp2)) # omc: (critical frequency that marks total internal reflection)
+                fm = 0.0
```

Comment:
Let's use `FT` to make sure that we have type-stable results for Float32. (here and everywhere in `src/`)